File: app/src/main/java/chat/rocket/android/layouthelper/chatroom/AbstractMessageViewHolder.java
package chat.rocket.android.layouthelper.chatroom;

import android.support.annotation.Nullable;
import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;

import chat.rocket.android.R;
import chat.rocket.android.helper.DateTime;
import chat.rocket.android.helper.TextUtils;
import chat.rocket.android.widget.AbsoluteUrl;
import chat.rocket.android.widget.RocketChatAvatar;
import chat.rocket.core.SyncState;

public abstract class AbstractMessageViewHolder extends ModelViewHolder<PairedMessage> {
  protected final RocketChatAvatar avatar;
  protected final ImageView errorImageView;
  protected final TextView username;
  protected final TextView subUsername;
  protected final TextView timestamp;
  protected final View userAndTimeContainer;
  protected final AbsoluteUrl absoluteUrl;
  protected final String hostname;
  protected final View newDayContainer;
  protected final TextView newDayText;

  /**
   * constructor WITH hostname.
   */
  public AbstractMessageViewHolder(View itemView, AbsoluteUrl absoluteUrl, String hostname) {
    super(itemView);
    avatar = itemView.findViewById(R.id.user_avatar);
    errorImageView = itemView.findViewById(R.id.errorImageView);
    username = itemView.findViewById(R.id.username);
    subUsername = itemView.findViewById(R.id.sub_username);
    timestamp = itemView.findViewById(R.id.timestamp);
    userAndTimeContainer = itemView.findViewById(R.id.user_and_timestamp_container);
    newDayContainer = itemView.findViewById(R.id.dayContainer);
    newDayText = itemView.findViewById(R.id.day);
    this.absoluteUrl = absoluteUrl;
    this.hostname = hostname;
  }

  /**
   * bind the view model.
   */
  public final void bind(PairedMessage pairedMessage, boolean autoloadImages) {
    if (pairedMessage.target.getSyncState() == SyncState.FAILED ||
        pairedMessage.target.getSyncState() == SyncState.DELETE_FAILED) {
      errorImageView.setVisibility(View.VISIBLE);
    } else {
      errorImageView.setVisibility(View.GONE);
    }

    bindMessage(pairedMessage, autoloadImages);
    renderNewDayAndSequential(pairedMessage);
  }

  protected abstract void bindMessage(PairedMessage pairedMessage, boolean autoloadImages);

  private void renderNewDayAndSequential(PairedMessage pairedMessage) {
    //see Rocket.Chat:packages/rocketchat-livechat/app/client/views/message.coffee
    if (!pairedMessage.hasSameDate()) {
      setNewDay(DateTime.fromEpocMs(pairedMessage.target.getTimestamp(), DateTime.Format.DATE));
      setSequential(false);
    } else if (!pairedMessage.target.isGroupable() || !pairedMessage.nextSibling.isGroupable()
        || !pairedMessage.hasSameUser()) {
      setNewDay(null);
      setSequential(false);
    } else {
      setNewDay(null);
      setSequential(true);
    }
  }

  private void setSequential(boolean sequential) {
    if (avatar != null) {
      if (sequential) {
        avatar.setVisibility(View.GONE);
      } else {
        avatar.setVisibility(View.VISIBLE);
      }
    }

    if (userAndTimeContainer != null) {
      if (sequential)
        userAndTimeContainer.setVisibility(View.GONE);
      else
        userAndTimeContainer.setVisibility(View.VISIBLE);
    }
  }

  private void setNewDay(@Nullable String text) {
    if (newDayContainer != null) {
      if (TextUtils.isEmpty(text)) {
        newDayContainer.setVisibility(View.GONE);
      } else {
        newDayText.setText(text);
        newDayContainer.setVisibility(View.VISIBLE);
      }
    }
  }
}

File: app/src/main/java/chat/rocket/android/layouthelper/chatroom/AbstractMessageViewHolder.kt
package chat.rocket.android.layouthelper.chatroom

import android.support.constraint.ConstraintLayout
import android.view.View
import android.widget.ImageView
import android.widget.TextView
import chat.rocket.android.helper.DateTime
import chat.rocket.android.widget.RocketChatAvatar
import chat.rocket.core.SyncState
import kotlinx.android.synthetic.main.day.view.*
import kotlinx.android.synthetic.main.item_room_message.view.*

abstract class AbstractMessageViewHolder(itemView: View, protected val hostname: String) : ModelViewHolder<PairedMessage>(itemView) {
    private val dayLayout: View by lazy { itemView.dayLayout }
    private val errorImage: ImageView by lazy { itemView.errorImage }
    private val middleContainer: ConstraintLayout by lazy { itemView.middleContainer }
    private val day: TextView by lazy { itemView.day }
    val avatar: RocketChatAvatar by lazy { itemView.avatar }
    val realName: TextView by lazy { itemView.realName }
    val username: TextView by lazy { itemView.username }
    val timestamp: TextView by lazy { itemView.timestamp }

    protected abstract fun bindMessage(pairedMessage: PairedMessage, autoLoadImage: Boolean)

    override fun bind(model: PairedMessage, autoLoadImages: Boolean) {
        if (model.target.syncState == SyncState.FAILED) {
            errorImage.visibility = View.VISIBLE
        } else {
            errorImage.visibility = View.GONE
        }

        bindMessage(model, autoLoadImages)
        renderSequentialAndNewDay(model)
    }

    private fun renderSequentialAndNewDay(pairedMessage: PairedMessage) {
        val target = pairedMessage.target
        val nextSibling = pairedMessage.nextSibling
        if (target != null && nextSibling != null) {
            if (target.isGroupable && nextSibling.isGroupable && pairedMessage.hasSameUser()) {
                setSequential(true)
            } else {
                setSequential(false)
            }
        }


        if (!pairedMessage.hasSameDate()) {
            showNewDay(DateTime.fromEpocMs(pairedMessage.target.timestamp, DateTime.Format.DATE))
        } else {
            showNewDay(null)
        }
    }

    private fun showNewDay(text: String?) {
        if (text.isNullOrEmpty()) {
            dayLayout.visibility = View.GONE
        } else {
            day.text = text
            dayLayout.visibility = View.VISIBLE
        }
    }

    private fun setSequential(showSequential: Boolean) {
        if (showSequential) {
            avatar.visibility = View.GONE
            middleContainer.visibility = View.GONE
        } else {
            avatar.visibility = View.VISIBLE
            middleContainer.visibility = View.VISIBLE
        }
    }
}